src/Microsoft.AspNetCore.Blazor.Server/Circuits/DefaultCircuitFactory.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Blazor.Browser.Rendering;
+using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    internal class DefaultCircuitFactory : CircuitFactory
+    {
+        private readonly IServiceScopeFactory _scopeFactory;
+
+        public DefaultCircuitFactory(IServiceScopeFactory scopeFactory)
+        {
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
+
+            StartupActions = new List<(PathString path, Action<BrowserRenderer> config)>();
+        }
+
+        public List<(PathString path, Action<BrowserRenderer> config)> StartupActions { get; }
+
+        public override CircuitHost CreateCircuit(HttpContext httpContext, IClientProxy client)
+        {
+            Action<BrowserRenderer> config = null;
+            for (var i = 0; i < StartupActions.Count; i++)
+            {
+                if (httpContext.Request.Path.StartsWithSegments(StartupActions[i].path))
+                {
+                    config = StartupActions[i].config;
```

Comment:
Should we be running *all* the startup actions for a given path, or just the last one?

I'm unsure what the use case would be for having multiple startup actions for a given path. If there is only to be one, it would be good to enforce that by storing them in a dictionary.

Reply:
This should be running one. We can make this a dictionary